### DISCUSSION TITLE
Not able to connect with ssl through proxy

### DISCUSSION THREAD

**Discussion by AndreAfonso87** [quickfix-j/quickfixj#755] [Q&A] [answered]
**Describe the bug**
We are trying to connect with SSL through Proxy, but the connection is not being made.

**To Reproduce**
Have set properties:
      ProxyType=http
      ProxyHost=XXX
      ProxyPort=80

**Expected behavior**
SSL connection through proxy

**system information:**
 - OS: RedHat 8.9
 - Java version 17
 - QFJ Version 2.3.1
 - SpringBoot 2.7.18
 - mina-core version: 2.1.8

**Additional context**
Startup logs:
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::               (v2.7.18)

{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"The following 1 profile is active: \"fix-adapter-APP-current\"","thread_name":"main","@timestamp":"2024-01-08T15:32:45.502+0000","level":"INFO","logger_name":"com.dummy.fix.FixAdapterRunner"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Started FixAdapterRunner in 14.723 seconds (JVM running for 19.232)","thread_name":"main","@timestamp":"2024-01-08T15:32:56.713+0000","level":"INFO","logger_name":"com.dummy.fix.FixAdapterRunner"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"IDLE State. Passive mode","thread_name":"main","@timestamp":"2024-01-08T15:32:56.727+0000","level":"INFO","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"JMS Queue browsing started for queue: UK.APP.FIXADAPTER.CONTROLLER","thread_name":"main","@timestamp":"2024-01-08T15:32:56.896+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Found messages: true","thread_name":"main","@timestamp":"2024-01-08T15:32:56.940+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Waiting 50000 milliseconds","thread_name":"main","@timestamp":"2024-01-08T15:32:56.940+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"JMS Queue browsing started for queue: UK.APP.FIXADAPTER.CONTROLLER","thread_name":"main","@timestamp":"2024-01-08T15:33:46.977+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Found messages: true","thread_name":"main","@timestamp":"2024-01-08T15:33:47.010+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Waiting 50000 milliseconds","thread_name":"main","@timestamp":"2024-01-08T15:33:47.011+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"JMS Queue browsing started for queue: UK.APP.FIXADAPTER.CONTROLLER","thread_name":"main","@timestamp":"2024-01-08T15:34:37.042+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Found messages: false","thread_name":"main","@timestamp":"2024-01-08T15:34:37.062+0000","level":"DEBUG","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"I'm now in a LIVE State","thread_name":"main","@timestamp":"2024-01-08T15:34:37.063+0000","level":"INFO","logger_name":"com.dummy.fix.AppController"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"[FIXT.1.1:SANUUATFIX->RTNSFIXUAT] monday, tuesday, wednesday, thursday, friday, saturday, 03:00:00-UTC - 21:55:00-UTC (monday, tuesday, wednesday, thursday, friday, saturday, 03:00:00-GMT - 21:55:00-GMT)","thread_name":"main","@timestamp":"2024-01-08T15:34:39.695+0000","level":"INFO","logger_name":"quickfix.DefaultSessionSchedule"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Session FIXT.1.1:SANUUATFIX->RTNSFIXUAT schedule is monday, tuesday, wednesday, thursday, friday, saturday, 03:00:00-UTC - 21:55:00-UTC (monday, tuesday, wednesday, thursday, friday, saturday, 03:00:00-GMT - 21:55:00-GMT)","thread_name":"main","@timestamp":"2024-01-08T15:34:40.032+0000","level":"INFO","logger_name":"quickfixj.event"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Created session: FIXT.1.1:SANUUATFIX->RTNSFIXUAT","thread_name":"main","@timestamp":"2024-01-08T15:34:40.033+0000","level":"INFO","logger_name":"quickfixj.event"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Socket option: SocketTcpNoDelay=true","thread_name":"main","@timestamp":"2024-01-08T15:34:40.053+0000","level":"INFO","logger_name":"quickfix.mina.NetworkingOptions"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Socket option: SocketSynchronousWrites=false","thread_name":"main","@timestamp":"2024-01-08T15:34:40.053+0000","level":"INFO","logger_name":"quickfix.mina.NetworkingOptions"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Socket option: SocketSynchronousWriteTimeout=30000","thread_name":"main","@timestamp":"2024-01-08T15:34:40.053+0000","level":"INFO","logger_name":"quickfix.mina.NetworkingOptions"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Configured socket addresses for session: [XXX.net/XXX.XX.XX.XXX:16003]","thread_name":"main","@timestamp":"2024-01-08T15:34:40.520+0000","level":"INFO","logger_name":"quickfixj.event"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"SessionTimer started","thread_name":"main","@timestamp":"2024-01-08T15:34:40.522+0000","level":"INFO","logger_name":"quickfix.SocketInitiator"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Started QFJ Message Processor","thread_name":"QFJ Message Processor","@timestamp":"2024-01-08T15:34:40.526+0000","level":"INFO","logger_name":"quickfix.SocketInitiator"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Pending connection not established after 4008 ms.","thread_name":"QFJ Timer","@timestamp":"2024-01-08T15:34:44.531+0000","level":"INFO","logger_name":"quickfixj.event"}
{"@version":1,"source_host":"fix-adapter-APP-current-43-qt7dm","message":"Pending connection not established after 9010 ms.","thread_name":"QFJ Timer","@timestamp":"2024-01-08T15:34:49.533+0000","level":"INFO","logger_name":"quickfixj.event"}

**Comment by chrjohn** [comment]
> Pending connection not established after ...

To me it looks like there is a network timeout. Could you double check the network connection?
Are you able to verify that the connection works without proxy?

**Reply by AndreAfonso87** [reply]
Yes the network connection through proxy works, i did a curl/telnet cmd and it connects. We also have other environments that connect without proxy just fine. But we have requirement for this particular environment it needs to be through proxy. I find it weird that in the startup logs there is no mention to using proxy, should there be?

**Reply by AndreAfonso87** [reply]
Also we checked the proxy and there is no trace that the application is trying to connect through the proxy

**Reply by chrjohn** [reply]
> Also we checked the proxy and there is no trace that the application is trying to connect through the proxy

How could there be a trace when the connection times out? ;)

Did you check if you need one of the other proxy options? https://rawgit.com/quickfix-j/quickfixj/master/quickfixj-core/src/main/doc/usermanual/usage/configuration.html

**Comment by AndreAfonso87** [comment]
Well, if it tried to connect through the proxy and failed afterwards, lets say like a firewall.... But its not like that its not even hitting the proxy.
The proxy doesn't need user or password, because we have other applications (different technology) that connect through proxy and doesn't have it either.
Is there any way to increase logging, to understand if its trying to use proxy even?

**Reply by chrjohn** [reply]
Haven't tested this but I could imagine that if you create a thread dump of your application then there should be an instance of class `InitiatorProxyIoHandler`.
BTW, could you please post your full config (if required replace CompIDs and IPs with XXX).

**Reply by AndreAfonso87** [reply]
Attached the thread dump
[jstack_threaddump.txt](https://github.com/quickfix-j/quickfixj/files/13906854/jstack_threaddump.txt)

**Reply by AndreAfonso87** [reply]
Here's the rest of the config: 
```
[DEFAULT]
JdbcSessionIdDefaultPropertyValue=appXXX-1

UseDataDictionary=Y
ValidateUnorderedGroupFields=N
ValidateUserDefinedFields=N
LogonTimeout=60
LogoutTimeout=20
SLF4JLogPrependSessionID=N
SLF4JLogHeartbeats=N

[SESSION]
ConnectionType=initiator
BeginString=FIXT.1.1
DefaultApplVerID=FIX.5.0SP1
TransportDataDictionary=appXXX/FIXT11.xml
AppDataDictionary=appXXX/FIX50SP1_TRTN.xml

ReconnectInterval=60
ResetOnLogon=Y
SenderCompID=XXX
TargetCompID=XXX

Weekdays=Monday,Tuesday,Wednesday,Thursday,Friday,Saturday
TimeZone=Europe/London
StartTime=03:00:00
EndTime=21:55:00


HeartBtInt=30
SocketConnectPort=16003
SocketConnectHost=pts.ppe.trading.appXXX.net
SocketUseSSL=Y
SocketKeyStore=/opt/XXX/certs/keystore/XXX.pfx
SocketKeyStorePassword=bdd6dd8a48ee4dbb1b45a5cc2d95917792db09eb24eaf2723afebb7fa720546c
ProxyType=http
ProxyHost=b2bproxy10g.XXX.corp
ProxyPort=8080
SocketLocalPort=5430
ChannelName=appXXX-FixAdapter
```

**Reply by chrjohn** [reply]
Sorry I meant a heap dump, not a thread dump, my fault.
Or get a class histogram with jmap, that will also list all used classes.

**Reply by AndreAfonso87** [reply]
Sorry for the delay, only managed to do it today. Attached the jmap histo file. I also have a dump, but its too big to annex here.
[Uploading jmap_histo.txt…]()

**Reply by AndreAfonso87** [reply]
[jmap_histo.txt](https://github.com/quickfix-j/quickfixj/files/13951622/jmap_histo.txt)

**Reply by chrjohn** [reply]
OK, so what do you want me to do now? ;) I think you can check for yourself if there are instances of InitiatorProxyIoHandler in the output. :)

BTW could you post the curl command that you use to verify that the connection through the proxy is working?

**Reply by AndreAfonso87** [reply]
I do not understand why its not working. We are calling:
SocketInitiator(fixApplication, storeFactory, sessionHandler.getSessionSettings(), logFactory, messageFactory);
Which if im seing this correctly it then calls createInitiator and inside its checking if it has properties ProxyType ProxyHost & ProxyPort
But its not working. i Guess its something to do with the properties file, but it seems fine i don't know what it could be.
Attached the curl cmd test with the proxy
![curl_tests](https://github.com/quickfix-j/quickfixj/assets/47325198/ca5fad14-a0f7-4c4d-add7-f97128c3d1d5)

**Reply by chrjohn** [reply]
You have `SocketLocalPort=5430` in your settings. Is there a reason for this? I can't see it specified in the curl command, so try to remove it from the settings. 
Also try setting `ProxyVersion=1.1`.
Moreover you are not using SSL in the curl command, so in the end you cannot really compare it to the QFJ setup.

**Comment by AndreAfonso87** [comment, marked as answer]
Fixed now. For future reference, using URL in SocketConnectionHost did not work. I had to open firewall specifically to the ip address and static ip address instead of url on SocketConnectionHost.

Thanks for your help

**Reply by chrjohn** [reply]
Glad that you managed to find the issue and thanks for sharing it here. 👍
